### PR TITLE
Add possibility to store in system permanently deleted users label...

### PR DESCRIPTION
**Pull request by adudekyeti**
Add possibility to store in system permanently deleted users label, also added possibility to update users label in admin menu.

### CODE REVIEW

File: modules/Vtiger/crons/LabelUpdater.php [hunk 1]
```diff
@@ -79,5 +79,13 @@ public function process()
...
 		}
 		$dataReader->close();
+		$usersRecordModel = Users_Record_Model::getAll(false);
+		foreach($usersRecordModel as $userRecordModel){
+			$userRecordModel->updateLabel();
+			--$limit;
+			if (0 === $limit) {
+				return;
+			}
+		}
```
Comment (rskrzypczak) [inline modules/Vtiger/crons/LabelUpdater.php]: for optimization

File: modules/Users/models/Record.php [hunk 2]
```diff
@@ -350,6 +350,7 @@ public function afterSaveToDb()
 			$dbCommand->update('vtiger_ossemployees', ['dav_status' => 1])->execute();
 		}
 		self::cleanCache($this->getId());
+		$this->deleteLabel();
```
Comment (rskrzypczak) [inline modules/Users/models/Record.php]: ?

**Review by mariuszkrzaczkowski — APPROVED** [review]

File: modules/Users/models/Record.php [hunk 3]
```diff
@@ -1042,4 +1043,59 @@ public function getFavouritesUsers()
...
+	 *
+	 * @return void
+	 */
+	public static function updateLabelUser(int $userId): void
+	{
+		if (\App\User::isExists($userId)) {
+			$userRecordModel = self::getInstanceById($userId, 'Users');
+			$userRecordModel->updateLabel();
+		}
+	}
```
Comment (rskrzypczak) [inline modules/Users/models/Record.php]: one static function
and maybe App\User or even Users_Model would be a better place

File: modules/Settings/Search/models/Module.php [hunk 4]
```diff
@@ -102,14 +102,20 @@ public static function save($params)
...
 	 */
-	public static function updateLabels($params)
+	public static function updateLabels($params): void
 	{
 		$moduleName = App\Module::getModuleName((int) $params['tabid']);
 		$db = App\Db::getInstance();
-		$db->createCommand()->update('u_#__crmentity_search_label', ['searchlabel' => ''], ['setype' => $moduleName])->execute();
-		$subQuery = (new \App\Db\Query())->select(['crmid'])->from('vtiger_crmentity')->where(['setype' => $moduleName]);
-		$db->createCommand()->delete('u_#__crmentity_label', ['crmid' => $subQuery])->execute();
+		if ('Users' !== $moduleName) {
```
Comment (rskrzypczak) [inline modules/Settings/Search/models/Module.php]: ?

File: modules/Users/models/Record.php (hunk 3 above)
Comment (mariuszkrzaczkowski) [reply modules/Users/models/Record.php]: App\User::updateAllLabels
App\User::updateUserLabel

File: app/User.php [hunk 5]
```diff
@@ -497,4 +499,29 @@ public static function getNumberOfUsers(): int
...
 	}
+
+	/**
+	 * Update users labels.
+	 *
+	 * @param int $fromUserId
+	 *
+	 * @return void
+	 */
+	public static function updateLabels(int $fromUserId = 0): void
```
Comment (rskrzypczak) [inline app/User.php:510]: where is $fromUserId ?

File: app/Fields/Owner.php [hunk 6]
```diff
@@ -634,22 +634,25 @@ public static function getUserLabel($id, $single = false)
...
+			}
+			$userLabel = isset($users[$id]) ? $users[$id]['fullName'] : false;
 		} else {
-			$instance = new self();
-			if ($single) {
-				$users = $instance->initUsers(false, $id);
-			} else {
-				$users = $instance->initUsers(false);
+			$users = \App\User::getAllUsersLabel();
+			if ($users) {
```
Comment (mariuszkrzaczkowski) [inline app/Fields/Owner.php]:
```suggestion
			if ($users = \App\User::getAllUsersLabel()) {
```

File: app/User.php [hunk 7]
```diff
@@ -350,28 +350,30 @@ public function isActive()
...
-			if (isset($users[$id]) && !$users[$id]['deleted']) {
+			if (($active && isset($users[$id]) && 'Active' == $users[$id]['status'] && !$users[$id]['deleted']) || (!$active && isset($users[$id]))) {
 				$isExists = true;
 			}
 		} else {
-			$isExists = (new \App\Db\Query())
-				->from('vtiger_users')
-				->where(['status' => 'Active', 'deleted' => 0, 'id' => $id])
-				->exists();
+			$isExists = (new \App\Db\Query())->from('vtiger_users')->where(['id' => $id]);
```
Comment (mariuszkrzaczkowski) [inline app/User.php]:
```suggestion
			$isExistsQuery = (new \App\Db\Query())->from('vtiger_users')->where(['id' => $id]);
```

File: app/User.php [hunk 8]
```diff
@@ -350,28 +350,30 @@ public function isActive()
...
+			if ($active) {
+				$isExists->andWhere(['status' => 'Active', 'deleted' => 0]);
```
Comment (mariuszkrzaczkowski) [inline app/User.php]:
```suggestion
				$isExistsQuery->andWhere(['status' => 'Active', 'deleted' => 0]);
```

File: modules/Vtiger/uitypes/Owner.php [hunk 9]
```diff
@@ -71,13 +71,17 @@ public function getDisplayValue($value, $record = false, $recordModel = false, $
...
 			case 'Users':
-				$userModel = Users_Privileges_Model::getInstanceById($value);
-				$userModel->setModule('Users');
-				if ('Inactive' === $userModel->get('status')) {
-					$ownerName = '<span class="redColor"><s>' . $ownerName . '</s></span>';
-				} elseif (\App\Privilege::isPermitted('Users', 'DetailView', $value)) {
-					$detailViewUrl = 'index.php?module=Users&view=Detail&record=' . $value;
-					$popoverRecordClass = 'class="js-popover-tooltip--record"';
+				if(!\App\User::isExists($value, false)){
+					$ownerName = '<span class="text-muted"><s>' . $ownerName . '</s></span>';
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/uitypes/Owner.php]:
```suggestion
					$ownerName = '<span class="text-muted"><s>' . $ownerName ?: '---' . '</s></span>';
```

**Pull request merged**